Qt: Remove unnecessary image buffer for Mac dock icon

droark opened the pull request:
The code curently writes the Mac dock icon to a buffer and then sets the final image data from the buffer. This was presumably done to satisfy Qt 4. Now that Qt 5 is mandatory, remove the buffer and write the data directly to the image.

Note that this patch will force the bitcoin-qt binary to have a minimum of Qt 5.2 on Macs due to usage of QtMacExtras (https://wiki.qt.io/New_Features_in_Qt_5.2). While this needs to be discussed before accepting the patch, Qt 5.2 is almost five years old. It's arguable that a Mac with Qt 5 but staying below 5.2 is an improbable corner case.

droark committed:
Remove unnecessary image buffer for Mac dock icon

The code curently writes the Mac dock icon to a buffer and then sets the final image data from the buffer. This was presumably done to satisfy Qt 4. Now that Qt 5 is mandatory, remove the buffer and write the data directly to the image.

Note that this patch will force the bitcoin-qt binary to have a minimum of Qt 5.2 on Macs due to usage of QtMacExtras (https://wiki.qt.io/New_Features_in_Qt_5.2). While this needs to be discussed before accepting the patch, Qt 5.2 is almost five years old. It's arguable that a Mac with Qt 5 but staying below 5.2 is an improbable corner case.

promag commented:
Discussion about minimum Qt5 version in #13478.

utACK ec16f4a.

promag commented:
Build error:
```
checking for QT5... no
configure: WARNING: Qt dependencies not found; bitcoin-qt frontend will not be built
```

fanquake commented:
This would be a nice to have cleanup/simplification. I've tested on macOS 10.13.5, building with Qt 5.11.1:

<img width="123" alt="macos-master" src="https://user-images.githubusercontent.com/863730/42131530-63e509a6-7d36-11e8-97db-7f0b859fb4d6.png">

ec16f4a:
<img width="119" alt="macos-13561" src="https://user-images.githubusercontent.com/863730/42131528-616dd4a0-7d36-11e8-843d-e3ce6644dfec.png">

However, there are a couple issues:

This PR isn't the place to discuss it, but I'm currently Concept NACK on adding per platform Qt minimums. I'll continue that discussion in #13478.

As pointed out [above](https://github.com/bitcoin/bitcoin/pull/13561#issuecomment-400986101), the macOS Travis build is currently failing. I've verified locally using a native depends build:
```
checking for Qt5Core Qt5Gui Qt5Network Qt5Widgets Qt5MacExtras... no
configure: WARNING: Qt dependencies not found; bitcoin-qt frontend will not be built
checking whether to build Bitcoin Core GUI... no (Qt5)
```
The issue is that we aren't currently including macExtras with qt in depends, so if we want to do this our qt package will need to be updated.

Also, if this change is going to force the macOS Qt minimum to be 5.2, can you drop this Qt VESION check:
https://github.com/bitcoin/bitcoin/blob/a6ed99a1e6facd38913711106bce6fb65bd14862/src/qt/macdockiconhandler.mm#L53

droark commented:
@fanquake - Thanks. Was looking into all that when you commented and was looking into the Qt package. I'll pick up the conversation at 13478 and update this PR as necessary.

jonasschnelli commented:
Concept ACK

fanquake commented:
@droark Have you had a chance to follow up here?

droark commented:
@fanquake - Hi. I started working on some depends stuff and then got distracted. I'll pick it up ASAP. I'll see if I can optimize any more Mac-specific code but that might take a little longer.

laanwj commented:
> Note that this patch will force the bitcoin-qt binary to have a minimum of Qt 5.2 on Macs due to usage of QtMacExtras

That's acceptable.

fanquake commented:
@droark Are you planning on continuing this? Otherwise II'll label as "Up for Grabs". \

The 5.2 requirement is no longer an issue, as master currently requires >5.4 anyways. The plan for 0.18 will likely be a minimum required version above 5.2 regardless.

droark commented:
@fanquake - Yes, I'm still on it. I was just waiting for a decision related to Qt. I'll rebase it right now and look into the failures from there.

droark committed:
Remove QtMacExtras requirement

Cross-compiling for platform-specific Qt extras may not be possible. Work around by copying the code from QtMac::toNSImage() and modifying as necessary. A redundant Qt version check was also removed.

droark commented:
Added a new commit that incorporated feedback. I also removed QtMacExtras. As best I can tell, trying to cross-compile Qt "extras" packages isn't a good idea. I did an end-run by just copying the code for QtMac::toNSImage() and modifying as necessary.

droark commented:
Hmmm. Still getting an error when cross-compiling. In particular, toCGImage() still requires QtMacExtras, and the underlying code seems to tap into a private Qt API that we have no business using. (Besides, it fails as-is.) This PR may need some rethinking. Anybody have any ideas? The cleanest thing to do would be to cross-compile QtMacExtras. I'm just not sure it's doable, or sane if it is doable. If not, this PR will probably have to be withdrawn, or radically reworked. The current method, while inefficient, manages to avoid Mac-specific calls.

droark commented:
@DrahtBot - Thanks. I'll look at that PR more closely. At a glance, it should be a cleaner merge, so I'll probably work around that one (assuming my PR can be salvaged, per my posts above).

fanquake commented:
@droark Given that #14597 removes the entire `macdockiconhandler.mm` file, I'm not sure there'll be much to salvage (code wise) here. I'm going to close this for now, if it turns out this is still needed, i.e 14597 falls through we'll reopen. If you've got further macOS depends related, please open a new PR.
